text/0000-private-mirrors.md

```diff
@@ -0,0 +1,146 @@
+
+# How it Works
+[how-it-works]: #how-it-works
+
+Out of scope - setting up the mirror itself. This is a separate concern that is not specific to Cloud Native Buildpacks.
+
+Once a mirror for one or more public registry has been setup, the platform operator can configure Cloud Native Buildpacks to use the mirror. This will permit `lifecycle` to use the mirrors for all images that are requested during builds using the `CNB_REGISTRY_MIRRORS` environment variable that would otherwise be requested from the public registry.
+
+The `CNB_REGISTRY_MIRRORS` environment variable will be a list of mirror configurations. Each mirror configuration will be a key/value pair, where the key is the registry URL and the value is the mirror URL. The key/value pairs will be separated by a semicolon (`;`).
+
+For example, if we wanted to configure a mirror for Docker Hub and Quay.io, we could set the `CNB_REGISTRY_MIRRORS` environment variable to the following:
+
+```
+docker.io=https://docker.mirror.example.com;quay.io=https://quay.mirror.example.com
+```
+
+When `lifecycle` requests an image during any phase (`analyze`, `restore`, `export`, `rebase`), it will first check the `CNB_REGISTRY_MIRRORS` environment variable. If the requested image's registry is configured in the `CNB_REGISTRY_MIRRORS` environment variable, it will use the mirror URL instead of the original registry URL.
+
+If the private registry requires authentication, authentication to the registry will be handled by the existing `CNB_REGISTRY_AUTH` value. If the private registry does not require authentication, no additional configuration is required.
+
+If registry mirrors are configured for specific images in configuration (e.g. `stack.toml` or `run.toml`), the `CNB_REGISTRY_MIRRORS` will be processed with each registry attempt. For example, the following `stack.toml` configures mirrors for `public/stack:run-image`:
+
+```toml
+[run-image]
+ image = "public/stack:run-image"
+ mirrors = ["quay.io/public/stack:run-image"]
+```
+
+If `CNB_REGISTRY_MIRRORS` has the value of:
+
+```
+docker.io=https://docker.mirror.example.com;quay.io=https://quay.mirror.example.com
+```
+
+When `lifecycle` attempts to resolve the `public/stack:run-image` image, `lifecycle` will attempt to fetch the image from `docker.mirror.example.com/public/stack:run-image`. If the [Run Image Resolution](https://github.com/buildpacks/spec/blob/main/platform.md#run-image-resolution) resulted in `quay.io/public/stack:run-image` being chosen, `lifecycle` will attempt to fetch the image from `quay.mirror.example.com/public/stack:run-image` instead. This new processing happens AFTER the [Run Image Resolution](https://github.com/buildpacks/spec/blob/main/platform.md#run-image-resolution) has executed.  The run image selection will NOT take `CNB_REGISTRY_MIRRORS` into account, but the final image resolution will. Think of `CNB_REGISTRY_MIRRORS` as a just-in-time override of the final image resolution.
```

Comment:
Just thinking out loud here, I'm not sure if this is relevant or not...

There exists today a `pack config run-image-mirrors` that will
* [transparently](https://github.com/buildpacks/pack/blob/3364a6b4b8b8aab03e90722f479a918f9dd3a8a9/pkg/client/build.go#L324) translate the run image reference

There is also a `pack config registry-mirrors` that will 
* [quietly](https://github.com/buildpacks/pack/blob/3364a6b4b8b8aab03e90722f479a918f9dd3a8a9/pkg/client/client.go#L218) translate the registry before fetching ANY image for ANY operation
* [transparently](https://github.com/buildpacks/pack/blob/3364a6b4b8b8aab03e90722f479a918f9dd3a8a9/pkg/client/build.go#L482) translate the registry for the run image reference AFTER `run-image-mirrors` have been applied

It's worth noting that while these translations update the `-run-image` provided to the lifecycle, they do NOT update stack.toml/run.toml i.e., these mirrors are "hidden" today in the sense that they are not persisted as metadata on the app image.

Reply:
Now I remember - it would be helpful to describe how `CNB_REGISTRY_MIRRORS` env var for the lifecycle would be exposed in pack given the above stuff that exists today. I would really hope that we could avoid adding yet another toggle here.

Reply:
@buildpacks/platform-maintainers @jjbustamante do you have any thoughts here?

Reply:
I tried to search in past issues and pull request to understand the reason for `pack config mirrors` commands. I found the following:

The order of preference applied for for run image when `pack build` is executed:

1. --run-image flag
2. default run image mirror, as discussed in the --publish/daemon cases above
    2.1 when `publish`, default to using the run-image mirror in the registry you are publishing the image to  
    2.2 when `daemon`, defaults to using the run-image mirror in the same registry (colocated) as the builder 
3. mirrors set using `pack config run-image-mirrors` 
4. mirrors defined on the builder

Based on this, the new `CNB_REGISTRY_MIRRORS` seems to be like a new option to override point **4**, right? something like:

4. expose some way to set `CNB_REGISTRY_MIRRORS` and pass it through the lifecycle
5. mirrors defined on the builder